### PR TITLE
Add API schema based request/response validation

### PR DESCRIPTION
**Pull request by ShalkiWenushika**
## Purpose
The WSO2 API Manager Microgateway currently does not have an approach to validate client requests that are sent to the Microgateway and also to validate responses coming from the backend. This feature will validate requests/responses based on a predefined API schema.

## Goals
Providing feature of enabling schema based request/response validation for product-microgateway. The user will be able to enable request validation, response validation or both request and response validation.

## Approach
1. Validating a request
- Sending a valid request
validrequest

- Sending an invalid request
invalid request

2.  Validating a response

- Backend sends a valid response
aa

- Backend sends an invalid resposponse
invalid response 1

         
Whenever a request comes to the microgateway that request should be validated against the swagger file of the requested API. If the request is valid it will be directed to the backend. If the request is invalid error 422 will be given to the user.

Whenever a response comes from the backend that response should be validated against the swagger file of the requested API. If the response is valid it will be directed to the user and if not error 500 will be given. Swagger file of the API will be embedded into the Microgateway at the time of microgateway setup.

The validation filter will check whether the user has enabled request/response validation and do the necessary validation. To enable this feature user has to edit the micro-gw.conf file. In the micro-gw.conf file there are two switches for request validation and response validation as enableRequestValidation and enableResponseValidation. So the user can enable request/response validation by making values of those variables into true. And also user needs to give the absolute path of the swagger file which is generated inside the SNAPSHOT to the variable, absolutePathToSwagger in the micro-gw.conf file.

## User stories

User Story | Microgateway with request validation enabled
-- | --
Overview | As a user, I only need to validate requests.
Persona | API creator, API Publisher
Prerequisites | <ul><li>User setup a Microgateway project</li><li>The user builds the Microgateway project</li><li>User goes into the built project, change the enableRequestValidation into true (enableRequestValidation=true) and give the absolute path of the swagger file which is generated inside the built project in to the variable, absolutePathToSwagger in the micro-gw.conf file</li><li>Run the Gateway</li></ul>
Narrative | <ul><li>The user sends a request to the microgateway</li><li> Microgateway will validate the request by getting the json payload of the request and comparing it with the relevant model defined in the swagger file</li><li>If the request is valid it will be directed to the backend. If the request is invalid, error with the HTTP status code 422 will be given to the user</li></ul>
Acceptance Criteria | Microgateway should send the request to the backend if only the request is valid. If not error 422 should be given to the user.

User Story | Microgateway with response validation enabled
-- | --
Overview | As a user, I only need to validate responses.
Persona | API creator, API Publisher
Prerequisites | <ul><li>User setup a Microgateway project</li><li>The user builds the Microgateway project</li><li>User goes into the built project, change the enableResponseValidation into true (enableResponseValidation=true) and give the absolute path of the swagger file which is generated inside the built project into the variable, absolutePathToSwagger in the micro-gw.conf file</li><li>Run the Gateway</li></ul>
Narrative | <ul><li>The user sends a request to the microgateway</li><li> Microgateway will validate the response comes from the backend by getting the json payload of the response and comparing it with the relevant model defined in the swagger file</li><li> If the response is valid it will be directed to the user</li><li> If the response is invalid, error with the HTTP status code 500 will be given to the user</li></ul>
Acceptance Criteria | Microgateway should send the response to the user if only the response is valid. If not error 500 should be given to the user.

User Story | Microgateway with request and response validation enabled
-- | --
Overview | As a user, I need to validate both requests and responses.
Persona | API creator, API Publisher
Prerequisites | <ul></li>User setup a Microgateway project</li><li>The user builds the Microgateway project</li><li>User goes into the built project, change the enableRequestValidation and enableResponseValidation into true (enableRequestValidation=true, enableResponseValidation=true) and give the absolute path of the swagger file which is generated inside the built project into the variable, absolutePathToSwagger in the micro-gw.conf file</li><li>Run the Gateway</li></ul>
Narrative | <ul><li>The user sends a request to the microgateway</li><li> Microgateway will validate the request by getting the json payload of the request and comparing it with the relevant model defined in the swagger file</li><li> If the request is valid it will be directed to the backend</li><li>If the request is invalid, error with the http status code 422 will be given to the user</li><li>If the request is valid, a response will come from the backend</li><li> Then the microgateway will validate that response</li><li>If the response is valid it will be directed to the user</li><li>If the response is invalid, error with the HTTP status code 500 will be given to the user.
Acceptance Criteria | Microgateway should send the request to the backend if only the request is valid. If not error 422 should be given to the user. Microgateway should send the response to the user if only the response is valid. If not error 500 should be given to the user.

## Automation tests
 - Integration tests
   ValidationTestCase

### CODE REVIEW

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/codegen/CodeGenerator.java [hunk 1]
```diff
@@ -159,6 +161,13 @@ private GenSrcFile generateCommonEndpoints() throws IOException {
     }
 
 
+    private GenSrcFile generateSwagger(BallerinaService context) throws IOException {
```
Comment (dinusha92) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/codegen/CodeGenerator.java]: Please add a method comment here.

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/rest/APIDetailedDTO.java [hunk 2]
```diff
@@ -417,7 +422,25 @@ public APICorsConfigurationDTO getCorsConfiguration() {
...
+    }
+
+    public void setApiSwagger(Json apiSwagger) {
+        this.apiSwagger = apiSwagger;
+    }
+
     public void setCorsConfiguration(APICorsConfigurationDTO corsConfiguration) {
         this.corsConfiguration = corsConfiguration;
     }
+
```
Comment (tharikaGitHub) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/rest/APIDetailedDTO.java]: Let's remove these 2 additional lines

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/rest/APIDetailedDTO.java [hunk 3]
```diff
@@ -417,7 +422,25 @@ public APICorsConfigurationDTO getCorsConfiguration() {
...
     }
 
+    public JsonObject getApiSwagger() {
+        String swagger = getApiDefinition();
+        JsonParser parser = new JsonParser();
+        JsonObject jsonSwagger;
+        try {
+            jsonSwagger = parser.parse(swagger).getAsJsonObject();
+        } catch (JsonSyntaxException e) {
+            throw new JsonSyntaxException("Error occured while parsing the swagger to a JsonObject");
```
Comment (dinusha92) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/rest/APIDetailedDTO.java]: We may need to add set the caught exception when throwing it again.

File: components/micro-gateway-core/src/main/ballerina/gateway/constants/api_security_constants.bal [hunk 4]
```diff
@@ -118,7 +124,11 @@ public function getAuthenticationFailureMessage(int errorCode) returns string {
         errorMessage = API_AUTH_INVALID_COOKIE_STRING_MESSAGE;
     } else if (errorCode == API_AUTH_NO_COOKIE_PROVIDED) {
         errorMessage = API_AUTH_NO_COOKIE_PROVIDED_STRING_MESSAGE;
-    } else {
+    } else if(errorCode == INVALID_ENTITY) {
```
Comment (tharikaGitHub) [inline components/micro-gateway-core/src/main/ballerina/gateway/constants/api_security_constants.bal]: Let's fix the formatting issues here

File: components/micro-gateway-core/src/main/ballerina/gateway/constants/api_security_constants.bal [hunk 5]
```diff
@@ -118,7 +124,11 @@ public function getAuthenticationFailureMessage(int errorCode) returns string {
...
+        errorMessage = INVALID_ENTITY_MESSAGE;
+    } else if(errorCode == INVALID_RESPONSE) {
+        errorMessage = INVALID_RESPONSE_MESSAGE;
+    }else {
```
Comment (dinusha92) [inline components/micro-gateway-core/src/main/ballerina/gateway/constants/api_security_constants.bal]: Please format the code here.

File: components/micro-gateway-core/src/main/ballerina/gateway/constants/constants.bal [hunk 6]
```diff
@@ -134,6 +134,8 @@
 @final public string THROTTLE_FILTER = "THROTTLE_FILTER";
 @final public string ANALYTICS_FILTER = "ANALYTICS_FILTER";
 @final public string MUTUAL_SSL_FILTER = "MUTUAL_SSL_FILTER";
+@final public string VALIDATION_FILTER = "VALIDATION_FILTER";
+
```
Comment (tharikaGitHub) [inline components/micro-gateway-core/src/main/ballerina/gateway/constants/constants.bal]: Remove this additional line

File: components/micro-gateway-core/src/main/ballerina/gateway/constants/constants.bal [hunk 7]
```diff
@@ -210,6 +214,66 @@
 @Description { value: "Destination" }
 @final public string DESTINATION = "destination";
 
+
```
Comment (tharikaGitHub) [inline components/micro-gateway-core/src/main/ballerina/gateway/constants/constants.bal]: Let's remove all the additional lines introduced in this file

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/ValidationTestCase.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (dinusha92) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/ValidationTestCase.java]: License header year should be 2019

File: components/micro-gateway-core/src/main/ballerina/gateway/constants/constants.bal [hunk 9]
```diff
@@ -210,6 +214,66 @@
...
+@final public string SWAGGER = "_swagger";
+@Description { value: "Seperator" }
+@final public string SEPERATOR = ".";
+@Description { value: "Relative path for the src folder" }
+@final public string BASEPATH = "../../../src/";
+
+
+
+
+
```
Comment (bhathiya) [inline components/micro-gateway-core/src/main/ballerina/gateway/constants/constants.bal]: extra new lines..

File: components/micro-gateway-core/src/main/ballerina/gateway/constants/constants.bal [hunk 10]
```diff
@@ -297,6 +361,16 @@
...
+@Description { value: "validation  related configs" }
+@final public string VALIDATION_CONFIG_INSTANCE_ID = "validationConfig";
+@Description { value: "Request validation enabled or not" }
+@final public string REQUEST_VALIDATION_ENABLED = "enableRequestValidation";
+@Description { value: "Response validation enabled or not" }
+@final public string RESPONSE_VALIDATION_ENABLED = "enableResponseValidation";
+@Description { value: "Absoulute path to swagger" }
+@final public string SWAGGER_ABSOLUTE_PATH = "absolutePathToSwagger";
+
+
```
Comment (bhathiya) [inline components/micro-gateway-core/src/main/ballerina/gateway/constants/constants.bal]: extra new line.

File: components/micro-gateway-core/src/main/ballerina/gateway/filters/validation_filter.bal [hunk 11]
```diff
@@ -0,0 +1,219 @@
+import ballerina/math;
```
Comment (pubudu538) [inline components/micro-gateway-core/src/main/ballerina/gateway/filters/validation_filter.bal]: Let's add the license header

**Pull request merged**